modules/ip4s/src/test/scala/pureconfig/module/ip4s/Ip4sTest.scala

```diff
@@ -1,12 +1,134 @@
 package pureconfig.module.ip4s
 
-import com.comcast.ip4s.{Hostname, Port}
+import scala.reflect.ClassTag
 
-import pureconfig.error.{CannotConvert, ConfigReaderFailures, ConvertFailure}
+import com.comcast.ip4s.Arbitraries._
+import com.comcast.ip4s._
+import com.typesafe.config._
+import org.scalacheck.{Arbitrary, Gen}
+import org.scalactic.source.Position
+import org.scalatest.Inside
+
+import pureconfig._
+import pureconfig.error._
 import pureconfig.syntax._
-import pureconfig.{BaseSuite, ConfigWriter}
 
-class Ip4sTest extends BaseSuite {
+class Ip4sTest extends BaseSuite with Inside {
+  // By default `minSuccessfull` set to 10 in `ScalaCheckDrivenPropertyChecks`.
+  // It is way too small for tests to be robust and reliable. Set to 100, which is used by default in ScalaCheck.
+  implicit override val generatorDrivenConfig = PropertyCheckConfiguration(minSuccessful = 100)
+
+  // This generator is missing from `com.comcast.ip4s.Arbitraries` for some reason.
+  private val hostGenerator: Gen[Host] =
+    Gen.oneOf(
+      ipGenerator,
+      hostnameGenerator,
+      // `idnGenerator` sometimes produces values that can be parsed as a valid `Hostname`.
+      // However, `Hostname` has a higher priority over `IDN` in `Host.fromString`.
+      // Therefore filter such values out.
+      idnGenerator.filter(idn => Hostname.fromString(idn.toString).isEmpty)
+    )
+
+  private def testsFor[A: ConfigWriter: ConfigReader, B: Arbitrary](
```

Comment:
This is quite a lot of custom test code - have you checked the utilities PureConfig provides in [ConfigConvertChecks](https://github.com/pureconfig/pureconfig/blob/master/testkit/src/main/scala/pureconfig/ConfigConvertChecks.scala) (available though `BaseSuite`)? Can we try to reuse them instead of having to maintain this alternative?

Reply:
Thank you for the hint – I've checked it out.

I found that I can make use of the `checkArbitrary` method in the tests and so did. However, it looks like there are no methods in `ConfigConvertChecks` that can help to verify failure scenarios, as well as verify that a particular `ConfigWriter` writes to an expected `ConfigValueType`.

So I re-used the `checkArbitrary` method in the tests but preserved checks for `ConfigValueType` and failure cases.

Let me know if you have concerns or objections about that please .